### PR TITLE
CAMS-582 Validation library

### PR DESCRIPTION
**Pull request by btposey**
# Purpose

Provide a common library of validation functions

# Major Changes

* Created a reusable, extendable library of validations functions
* Created a pattern for adding new validation functions
* Created a pattern for defining validation specs.

# Testing/Validation

Unit tests cover each validation function.

## Summary by Sourcery

Introduce a reusable, spec-driven validation framework and adopt it across both front-end and back-end trustee workflows.

New Features:
- Add a common validation library with spec definitions and core validator functions (minLength, matches, isInSet, optional, nullable, nested specs, etc.)
- Integrate spec-based validation into useTrusteeFormValidation hook and TrusteeCreateForm, replacing manual validation logic
- Apply validation specs in TrusteesUseCase on the backend for input sanitization and throw structured BadRequestError on failures
- Enhance ComboBox and UsStatesComboBox components to display field-level error messages

Enhancements:
- Introduce ZIP_REGEX and TRUSTEE_STATUS_VALUES constants for consistent pattern and status validation
- Refactor header to adjust the placement of the Trustees navigation link

Tests:
- Add extensive unit tests covering the validation library core and validator functions
- Update and streamline form validation tests and remove obsolete TrusteeCreateForm integration tests

### CODE REVIEW

File: common/src/cams/validation.ts [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+  };
+}
+
+function isEmailAddress(value: unknown): ValidatorResult {
+  return matches(/^[^\s@]+@[^\s@]+\.[^\s@]+$/, 'Must be a valid email address')(value);
+}
+
+function isPhoneNumber(value: unknown): ValidatorResult {
+  return matches(/^\d{10}$/, 'Must be a valid phone number')(value);
+}
```
[the remark this answers is not in the document]
Comment (btposey) [reply common/src/cams/validation.ts:69]: Our application is specifically limited to 10-digit North American number plan phone numbers.

File: common/src/cams/validation.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+export type ValidationSpec<T> = Record<keyof T, ValidatorFunction[]>;
+
+// Need to support dates, numbers. Nullable and undefined.
+// Required vs optional.
+
+function isString(value: unknown): ValidatorResult {
+  return typeof value === 'string' ? { valid: true } : { valid: false, reason: 'Must be a string' };
+}
+
+function minLength(length: number): ValidatorFunction {
```
[the remark this answers is not in the document]
Comment (btposey) [reply common/src/cams/validation.ts:12]: Good suggestions.

File: common/src/cams/validation.test.ts [hunk 3]
```diff
@@ -0,0 +1,637 @@
...
+    ];
+    test.each(testCases)('$description', (testCase) => {
+      expect(V.isString(testCase.value)).toEqual(testCase.expected);
+    });
+  });
+  describe('minLength', () => {
+    const testCases = [
+      {
+        description: 'should return valid for string meeting minimum length',
+        minLength: 3,
```
Comment (jamesobrooks) [inline common/src/cams/validation.test.ts:45]: I would rather this test a maximally shorter word than minimum.

```suggestion
        minLength: 4,
```

File: common/src/cams/validation.ts [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+
+function maxLength(length: number): ValidatorFunction {
+  return (value: unknown): ValidatorResult => {
+    return typeof value === 'string' && value.length <= length
+      ? { valid: true }
+      : { valid: false, reason: `Must be no more than ${length} characters long` };
+  };
+}
+
+function length(min: number, max: number, reason?: string): ValidatorFunction {
```
[the remark this answers is not in the document]
Comment (btposey) [reply common/src/cams/validation.ts:95]: Fixed.

**Comment by btposey** [comment]
@sourcery-ai review

File: common/src/cams/validation.ts [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+export type ValidatorFunction = (value: unknown) => ValidatorResult;
+export type ValidationSpec<T> = Record<keyof T, ValidatorFunction[]>;
+
+/**
+ * Validate a value against a validator function.
+ *
+ * @param func
+ * @param value
+ */
+function validate(func: ValidatorFunction, value: unknown): ValidatorResult {
```
[the remark this answers is not in the document]
Comment (btposey) [reply common/src/cams/validation.ts:20]: I disagree. I want to make these functions available as they are suitable to known use cases.

File: common/src/cams/validation.test.ts [hunk 6]
```diff
@@ -0,0 +1,682 @@
...
+        expected: { valid: false, reasons: ['Must match the pattern /^[a-z]+$/'] },
+      },
+    ];
+    test.each(testCases)('$description', (testCase) => {
+      const validator = V.matches(testCase.regex, testCase.error);
+      expect(validator(testCase.value)).toEqual(testCase.expected);
+    });
+  });
+
+  describe('isEmailAddress', () => {
```
Comment (En-8) [inline common/src/cams/validation.test.ts:275]: Given our chat last week about the differences in our regex patterns to catch edge-case-but-still-technically-valid email addresses, I think it would be beneficial to include several more test cases here covering some examples.
This way the test serves as documentation that "yes, we know these are allowed/disallowed, and we are good with that" for anyone that may come along and look at this sometime down the line.

Some examples:
```
// These are all valid according to the regex in this PR
.user@example.com
user.@example.com
user..name@example.com
user,name@example.com
user:name@example.com
user(name)@example.com
"user"@example.com
user@-example.com
user@example-.com
user@example..com
user@.example.com
user@exa_mple.com
user@exämple.com
Jörg@example.com
user@[127.0.0.1]
user@example/.com
?()!!&@?.?!
(comment)test@example.com

// This is invalid according to the regex in this PR
"why would anyone ever do this"@example.com
```

File: common/src/cams/validation.test.ts [hunk 7]
```diff
@@ -0,0 +1,682 @@
...
+      },
+      {
+        description: 'should return invalid for phone number with more than 10 digits',
+        value: '12345678901',
+        expected: { valid: false, reasons: ['Must be a valid phone number'] },
+      },
+      {
+        description: 'should return invalid for phone number with dashes',
+        value: '123-456-7890',
+        expected: { valid: false, reasons: ['Must be a valid phone number'] },
```
Comment (En-8) [inline common/src/cams/validation.test.ts:353]: If we're not going to allow common phone number formatting in our validation function, how do intend to still provide a UX that matches typical expectations for phone numbers in form inputs? Something like only allowing typing digits while the UI manages display format?

File: common/src/cams/validation.ts [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+}
+
+/**
+ * Validates whether a value is a valid email address format.
+ *
+ * @param {unknown} value - The value to be validated as an email address
+ * @returns {ValidatorResult} Object containing validation status and reason for failure if invalid
+ */
+function isEmailAddress(value: unknown): ValidatorResult {
+  return matches(/^[^\s@]+@[^\s@]+\.[^\s@]+$/, 'Must be a valid email address')(value);
```
Comment (En-8) [inline common/src/cams/validation.ts:229]: I don't normally suggest comments, but in this case since we've already had discussions around what regex to use, I think it would be beneficial to include a brief comment about _why_ we're using this one and not something more strict.

File: common/src/cams/validation.test.ts (hunk 6 above)
Comment (btposey) [reply common/src/cams/validation.test.ts:275]: Can do once we consolidate the trustee validation function onto this library.

File: common/src/cams/validation.test.ts (hunk 7 above)
Comment (btposey) [reply common/src/cams/validation.test.ts:353]: I would prefer allowing the UI to do all the formatting, but we can all discuss.

File: common/src/cams/validation.ts (hunk 8 above)
Comment (btposey) [reply common/src/cams/validation.ts:229]: Not a bad idea to track it in the JSDoc.

File: common/src/cams/validation.ts [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+/**
+ * Validates all properties of an object against a validation specification.
+ * Handles both validator function arrays and nested validation specs.
+ *
+ * @template T - The type of the object being validated
+ * @param spec - The validation specification defining rules for all object properties
+ * @param obj - The object to be validated
+ * @returns ValidatorResultsSet object containing validation status and property-specific reasons for failure
+ */
+export function validateObject(spec: ValidationSpec<unknown>, obj: unknown): ValidatorResult {
```
[the remark this answers is not in the document]
Comment (btposey) [reply common/src/cams/validation.ts:84]: We need the `validateKey` function when we want to validate only a single key from the spec so we cannot inline it into `validateObject`. Likewise we have uses for the `flattenReasonMap` separate from `flatten` so those functions cannot be consolidated either.

File: common/src/cams/validators.test.ts [hunk 10]
```diff
@@ -0,0 +1,757 @@
...
+        expected: { reason: 'Must contain at least 10 characters' },
+      },
+      {
+        description: 'should return invalid with multiple reasons when multiple validators fail',
+        validators: [Validators.minLength(10), Validators.maxLength(3)],
+        value: 'hello',
+        expected: {
+          reason: 'Must contain at least 10 characters',
+        },
+      },
```
[the remark this answers is not in the document]
Comment (btposey) [reply common/src/cams/validators.test.ts:499]: We only "short circuit" to the first reason in `validateEach`. As such the rest of the codebase is still capable of returning multiple reasons.

File: user-interface/src/trustees/UseTrusteeFormValidation.ts [hunk 11]
```diff
@@ -1,69 +1,42 @@
...
+  validateObject,
+} from '@common/cams/validation';
+
+const trusteeFormDataSpec: ValidationSpec<TrusteeFormData> = {
+  name: [V.minLength(1, 'Trustee name is required')],
+  address1: [V.minLength(1, 'Address is required')],
+  address2: [V.optional(V.maxLength(50))],
+  city: [V.minLength(1, 'City is required')],
+  state: [V.exactLength(2, 'State is required')],
+  zipCode: [V.matches(ZIP_REGEX, 'ZIP code must be 5 digits or 9 digits with a hyphen')],
```
[the remark this answers is not in the document]
Comment (btposey) [reply user-interface/src/trustees/UseTrusteeFormValidation.ts:19]: @sourcery-ai

File: user-interface/src/trustees/UseTrusteeFormValidation.ts [hunk 12]
```diff
@@ -122,22 +95,16 @@ export function useTrusteeFormValidation(): TrusteeFormValidation {
    * Checks if form is both valid (no errors) and complete (all required fields filled)
    */
   const isFormValidAndComplete = (formData: TrusteeFormData): boolean => {
```
[the remark this answers is not in the document]
Comment (btposey) [reply user-interface/src/trustees/UseTrusteeFormValidation.ts:97]: @sourcery-ai

File: common/src/cams/validators.ts [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+/**
+ * Creates a validator function that treats undefined values as valid and applies other validators otherwise.
+ * This allows for optional fields in validation schemas where undefined values are acceptable.
+ *
+ * @param {...ValidatorFunction[]} validators - Variable number of validator functions to apply when value is not undefined
+ * @returns {ValidatorFunction} A validator function that allows undefined values and validates defined values
+ */
+function optional(...validators: ValidatorFunction[]): ValidatorFunction {
```
[the remark this answers is not in the document]
Comment (btposey) [reply common/src/cams/validators.ts:27]: The skip-if-predicate is a good idea and I have implemented that. Otherwise I want to leave the length functions as they are. I do not mind having the dynamic reason text consolidated into a single function for reason strings that are otherwise going to be very redundant.

File: user-interface/src/trustees/UseTrusteeFormValidation.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (btposey) [reply user-interface/src/trustees/UseTrusteeFormValidation.ts:19]: @sourcery-ai Agreed, but we are going to use the REGEX to return a single reason for validation of this field.

File: user-interface/src/trustees/TrusteeCreateForm.test.tsx [hunk 14]
```diff
@@ -802,73 +776,6 @@ describe('TrusteeCreateForm', () => {
       } as unknown as ReturnType<typeof UseApi2Module.useApi2>);
     });
 
-    test('submit button starts disabled and enables only when form is valid and complete', async () => {
```
[the remark this answers is not in the document]
Comment (btposey) [reply user-interface/src/trustees/TrusteeCreateForm.test.tsx:805]: Done

File: user-interface/src/trustees/TrusteeCreateForm.tsx [hunk 15]
```diff
@@ -194,8 +194,10 @@ export default function TrusteeCreateForm() {
   }
 
   async function handleSubmit(ev: React.FormEvent) {
-    ev.preventDefault();
-    await submit();
+    if (isFormValidAndComplete(getFormData())) {
+      ev.preventDefault();
```
[the remark this answers is not in the document]
Comment (btposey) [reply user-interface/src/trustees/TrusteeCreateForm.tsx:198]: @sourcery-ai We do provide immediate feedback to the user when they click the submit button by flagging each error with each form field that failed validation. We purposefully leave the submit button activated.
